File: implicit/src/shaders/polygon.c

__kernel void apply_no_sign(
    __global float *buffer,
    ulong width,
    __global float *lines,
    ulong count,
    float m11, float m12,
    float m21, float m22,
    float m31, float m32
)
{
    size_t x = get_global_id(0);
    size_t y = get_global_id(1);
    size_t pos = x + y * width;

    float x_s = (float)x * m11 + (float)y * m21 + m31;
    float y_s = (float)x * m12 + (float)y * m22 + m32;

    if (count < 2)
    {
        buffer[pos] = NAN;
        return;
    }

    float minimum = INFINITY;
    int winding = 0;

    for (size_t i = 0; i < count; i += 4)
    {
        float x1 = lines[i + 0];
        float y1 = lines[i + 1];
        float x2 = lines[i + 2];
        float y2 = lines[i + 3];

        if (x1 == x2 && y1 == y2) {
            continue;
        }

        if (isnan(x1) ||
            isnan(x2) ||
            isnan(y1) ||
            isnan(y2)) {
            break;
        }

        float2 res = dist_to_line_comp(x_s, y_s, x1, y1, x2, y2);
        float new = res.x;
        float is_left = res.y;
        if (y1 <= y_s) {
            if (y2 > y_s) {
                if (is_left > 0.0) {
                    winding++;
                }
            }
        } else {
            if (y2 <= y_s) {
                if (is_left < 0.0) {
                    winding--;
                }
            }
        }
        minimum = min(minimum, fabs(new));
    }
    float s;
    if (winding == 0) {
        s = 1.0;
    } else {
        s = -1.0;
    }
    buffer[pos] = copysign(minimum, s);
}

__kernel void apply_with_sign(
    __global float *buffer,
    __global float *signbuffer,
    ulong width,
    __global float *lines,
    ulong count,
    float m11, float m12,
    float m21, float m22,
    float m31, float m32
)
{
    size_t x = get_global_id(0);
    size_t y = get_global_id(1);
    size_t pos = x + y * width;

    float x_s = (float)x;
    float y_s = (float)y;

    if (count < 2)
    {
        buffer[pos] = NAN;
        return;
    }

    float minimum_abs = INFINITY;

    for (size_t i = 0; i < count; i += 4)
    {
        float x1 = lines[i + 0];
        float y1 = lines[i + 1];
        float x2 = lines[i + 2];
        float y2 = lines[i + 3];

        if (x1 == x2 && y1 == y2) {
            continue;
        }

        if (isnan(x1) ||
            isnan(x2) ||
            isnan(y1) ||
            isnan(y2)) {
            break;
        }

        float new = dist_to_line(x_s, y_s, x1, y1, x2, y2);
        float new_abs = fabs(new);

        if (new_abs < minimum_abs) {
            minimum_abs = new_abs;
        }
    }

    buffer[pos] = copysign(minimum_abs, signbuffer[pos]);
}

File: core/implicit/src/polygon.rs
use geometry::Point;
use extern_api::Matrix;
use ocl::Kernel;
use opencl::{FieldBuffer, LineBuffer, OpenClContext};

const PROGRAM: &'static str = concat!(
    include_str!("shaders/dist_to_line.c"),
    include_str!("shaders/polygon.c")
);

pub fn run_poly<I>(
    points: I,
    signfield: Option<&mut FieldBuffer>,
    width: u32,
    height: u32,
    matrix: Matrix,
    ctx: &OpenClContext,
) -> Option<FieldBuffer>
where
    I: IntoIterator<Item = Point>,
{
    if let Some(signfield) = &signfield {
        assert_eq!(signfield.height, height);
        assert_eq!(signfield.width, width);
    }

    let _guard = ::flame::start_guard("run_poly");

    let mut buffer = vec![];
    for Point { x, y, .. } in points {
        buffer.push(x);
        buffer.push(y);
    }

    if buffer.len() == 0 {
        return None;
    }

    let buffer = ctx.line_buffer(&buffer[..]);
    let buffer_len = buffer.size();

    match signfield {
        Some(sf) => Some(run_poly_raw_with_sign(
            buffer, sf, width, height, buffer_len, matrix, ctx,
        )),
        None => Some(run_poly_raw_no_sign(buffer, width, height, matrix, ctx)),
    }
}

#[inline(always)]
pub fn add_matrix(kernel: Kernel, matrix: Matrix) -> Kernel {
    let matrix = matrix.inverse().unwrap();
    kernel.set_arg("m11", matrix.m11).unwrap();
    kernel.set_arg("m12", matrix.m12).unwrap();
    kernel.set_arg("m21", matrix.m21).unwrap();
    kernel.set_arg("m22", matrix.m22).unwrap();
    kernel.set_arg("m31", matrix.m31).unwrap();
    kernel.set_arg("m32", matrix.m32).unwrap();
    kernel
}

pub fn run_poly_raw_no_sign(
    lines: LineBuffer,
    width: u32,
    height: u32,
    matrix: Matrix,
    ctx: &OpenClContext,
) -> FieldBuffer {
    let _guard = ::flame::start_guard("run_poly_raw");
    let mut out = ctx.field_buffer(width, height, 1, None);
    let mut kernel = ctx.compile("apply_no_sign", PROGRAM, |register| {
        register.buffer("buffer");
        register.long("width");
        register.buffer("lines");
        register.long("count");
        register.matrix();
    });

    kernel.set_default_global_work_size(::ocl::SpatialDims::Two(width as usize, height as usize));
    kernel
        .set_arg("buffer", out.to_opencl(ctx.queue()))
        .unwrap();
    kernel.set_arg("width", width as u64).unwrap();
    kernel.set_arg("lines", lines.buffer()).unwrap();
    kernel.set_arg("count", lines.size()).unwrap();
    let kernel = add_matrix(kernel, matrix);
    unsafe {
        kernel.enq().unwrap();
    }
    out
}

pub fn run_poly_raw_with_sign(
    lines: LineBuffer,
    signfield: &mut FieldBuffer,
    width: u32,
    height: u32,
    count: usize,
    matrix: Matrix,
    ctx: &OpenClContext,
) -> FieldBuffer {
    let _guard = ::flame::start_guard("run_poly_raw");
    let mut out = ctx.field_buffer(width, height, 1, None);
    let mut kernel = ctx.compile("apply_with_sign", PROGRAM, |register| {
        register.buffer("buffer");
        register.buffer("signbuffer");
        register.long("width");
        register.buffer("lines");
        register.long("count");
        register.matrix();
    });

    kernel.set_default_global_work_size(::ocl::SpatialDims::Two(width as usize, height as usize));
    kernel
        .set_arg("buffer", out.to_opencl(ctx.queue()))
        .unwrap();
    kernel
        .set_arg("signbuffer", signfield.to_opencl(ctx.queue()))
        .unwrap();
    kernel.set_arg("width", width as u64).unwrap();
    kernel.set_arg("lines", lines.buffer()).unwrap();
    kernel.set_arg("count", count).unwrap();
    let kernel = add_matrix(kernel, matrix);
    unsafe {
        kernel.enq().unwrap();
    }
    out
}
